benches/bin/mvcctxn.rs

```diff
@@ -0,0 +1,364 @@
+extern crate rand;
+
+use std::time::SystemTime;
+use std::sync::mpsc::channel;
+
+use tikv::storage::{new_local_engine_with_config, Engine, Key, Modify, Mutation, Options,
+                    SnapshotStore, Statistics, ALL_CFS, TEMP_DIR};
+use tikv::config::DbConfig;
+use tikv::storage::mvcc::MvccTxn;
+use tikv::util::threadpool::{DefaultContext, ThreadPoolBuilder};
+use tikv::util::config::ReadableSize;
+use kvproto::kvrpcpb::{Context, IsolationLevel};
+
+use super::print_result;
+use test::BenchSamples;
+
+use rand::Rng;
+
+use utils::*;
+
+
+
+#[inline]
+fn do_write(engine: &Engine, modifies: Vec<Modify>) {
+    engine.write(&Context::new(), modifies).unwrap();
+}
+
+#[inline]
+fn prewrite(engine: &Engine, mutations: &[Mutation], primary: &[u8], start_ts: u64) {
+    let snapshot = engine.snapshot(&Context::new()).unwrap();
+    let mut txn = MvccTxn::new(snapshot, start_ts, None, IsolationLevel::SI, false);
+    for m in mutations {
+        txn.prewrite(m.clone(), primary, &Options::default())
+            .unwrap();
+    }
+    do_write(engine, txn.into_modifies());
+}
+
+#[inline]
+fn commit(engine: &Engine, keys: &[Key], start_ts: u64, commit_ts: u64) {
+    let snapshot = engine.snapshot(&Context::new()).unwrap();
+    let mut txn = MvccTxn::new(snapshot, start_ts, None, IsolationLevel::SI, false);
+    for key in keys {
+        txn.commit(key, commit_ts).unwrap();
+    }
+    do_write(engine, txn.into_modifies());
+}
+
+fn prepare_test_engine(versions: usize, value_len: usize, keys: &[Vec<u8>]) -> Box<Engine> {
+    let mut config = DbConfig::default();
+    config.defaultcf.write_buffer_size = ReadableSize::gb(1);
+    config.writecf.write_buffer_size = ReadableSize::gb(1);
+    config.lockcf.write_buffer_size = ReadableSize::gb(1);
+
+    let engine = new_local_engine_with_config(TEMP_DIR, ALL_CFS, &config, false).unwrap();
+
+    for _ in 0..versions {
+        for key in keys {
+            let value = vec![0u8; value_len];
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            prewrite(
+                &*engine,
+                &[Mutation::Put((Key::from_raw(key), value))],
+                key,
+                start_ts,
+            );
+            commit(&*engine, &[Key::from_raw(key)], start_ts, commit_ts);
+        }
+    }
+    engine
+}
+
+#[inline]
+fn get(engine: &Engine, key: &Key, statistics: &mut Statistics) -> Option<Vec<u8>> {
+    let snapshot = engine.snapshot(&Context::new()).unwrap();
+    let start_ts = next_ts();
+    let snapstore = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, false);
+    snapstore.get(key, statistics).unwrap()
+}
+
+fn bench_get(engine: &Engine, keys: &[Vec<u8>]) -> BenchSamples {
```

Comment:
For get, we should use an appropriate block-cache-size and turn `fill_cache` on.